File: solutions/csharp/robot-simulator/1/RobotSimulator.cs
using System;

public enum Direction
{
    North,
    East,
    South,
    West
}
public class RobotSimulator
{
    Direction direction;
    int x, y;
    public RobotSimulator(Direction direction, int x, int y)
    {
        this.direction = direction;
        this.x = x;
        this.y = y;
    }
    public Direction Direction
    {
        get
        {
            return direction;
        }
    }
    public int X
    {
        get
        {
            return x;
        }
    }
    public int Y
    {
        get
        {
            return y;
        }
    }
    public void Move(string instructions)
    {
        foreach (var step in instructions)
        {
            switch (step)
            {
                case 'A': Advance(); break;
                case 'R': TurnRight(); break;
                case 'L': TurnLeft(); break;
            }
        }
    }
    void Advance()
    {
        switch (direction)
        {
            case Direction.North:
                y++;
                break;
            case Direction.South:
                y--;
                break;
            case Direction.East:
                x++;
                break;
            case Direction.West:
                x--;
                break;
        }
    }
    void TurnRight()
    {
        direction++;
        if ((int)direction > 3) direction = (Direction)0;
    }
    void TurnLeft()
    {
        direction--;
        if ((int)direction < 0) direction = (Direction)3;
    }
}

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"

namespace robot_simulator {

// TODO: add your solution here
void Robot::turn_right() { bearing_ = static_cast<Bearing>((static_cast<int>(bearing_) + 1) % NUM_BEARING); }
void Robot::turn_left()
{
    bearing_ = static_cast<Bearing>((static_cast<int>(bearing_) - 1 + NUM_BEARING) % NUM_BEARING);
}
void Robot::advance()
{
    auto &[x, y] = pos_;
    switch (bearing_) {
    case Bearing::NORTH: {
        ++y;
    } break;
    case Bearing::EAST: {
        ++x;
    } break;
    case Bearing::SOUTH: {
        --y;
    } break;
    case Bearing::WEST: {
        --x;
    } break;
    }
}
void Robot::execute_sequence(const std::string &sequence)
{
    for (char order : sequence) {
        switch (order) {
        case 'A': {
            advance();
        } break;
        case 'R': {
            turn_right();
        } break;
        case 'L': {
            turn_left();
        } break;
        }
    }
}

} // namespace robot_simulator





